### PR TITLE
Added Wix installer native project.

### LINKED ISSUE #1276: Modifying an existing install via MSI doesn't install/uninstall any changed drivers
### Describe the bug

Modifying an existing install via MSI doesn't install/uninstall any changed drivers

### OS information

_No response_

### Steps taken to reproduce bug

* Install from MSI, leaving only the default (runtime) selected
* Run the MSI again, selecting Change, and add the "testing" component

### Expected behavior

* "sc query sampleebpfext" should show the driver running

### Actual outcome

* "sc query sampleebpfext" does not find the service
* "sc qc sampleebpfext" does not show the driver as installed

### Additional details

It is correctly installed if "testing" component was added at initial install time, but "Change" does not trigger install/uninstall of drivers.

### LINKED ISSUE #1412: MSI installer should install services with AutoStart instead of on OnDemand
### Describe the feature you'd like supported

The ebpf services are installed with `start=demand` which means if the machine reboots it they don't start back up. I encountered this after installing and rebooting due to a needed update (https://github.com/microsoft/ebpf-for-windows-demo/issues/22#issuecomment-1254266031)

### Proposed solution

Switch services to `start=auto` which mean if machine reboots ebpf starts up. I believe this would be expected behavior in most scenarios.

### Additional context

change would be

https://github.com/microsoft/ebpf-for-windows/blob/ecd08ecbb992be596090a473970afcc1eb815576/scripts/install_ebpf.psm1#L63

and maybe modify the install extension in https://github.com/microsoft/ebpf-for-windows/blob/ecd08ecbb992be596090a473970afcc1eb815576/scripts/install_ebpf.psm1#L73

### LINKED ISSUE #1478: MSI fails to install
### Describe the bug

Attempt to install the MSI

### OS information

_No response_

### Steps taken to reproduce bug

Start MSI to install. Leave default options.

### Expected behavior

Install succeeds.

### Actual outcome

Install fails. 

[install.log](https://github.com/microsoft/ebpf-for-windows/files/9824750/install.log)

```
WixQuietExec64:  WorkingDirectory is C:\Program Files\ebpf-for-windows\scripts\..
WixQuietExec64:  LogFileName is TestLog.log
WixQuietExec64:  Installing eBPF...
WixQuietExec64:   ::  OpenService FAILED 1060:
WixQuietExec64:   :: The specified service does not exist as an installed service.
WixQuietExec64:   ::  OpenService FAILED 1060:
WixQuietExec64:   :: The specified service does not exist as an installed service.
WixQuietExec64:   ::  OpenService FAILED 1060:
WixQuietExec64:   :: The specified service does not exist as an installed service.
WixQuietExec64:   ::  OpenService FAILED 1060:
WixQuietExec64:   :: The specified service does not exist as an installed service.
WixQuietExec64:   :: Element not found.
WixQuietExec64:   :: Installing NetEbpfExt...
WixQuietExec64:   ::  CreateService SUCCESS
WixQuietExec64:   :: NetEbpfExt driver created.
WixQuietExec64:   :: Installing EbpfCore...
WixQuietExec64:   ::  CreateService SUCCESS
WixQuietExec64:   :: EbpfCore driver created.
WixQuietExec64:  Failed to create eBPF user mode service.
WixQuietExec64:  At C:\Program Files\ebpf-for-windows\scripts\install_ebpf.psm1:75 char:9
WixQuietExec64:  +         throw ("Failed to create eBPF user mode service.")
WixQuietExec64:  +         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
WixQuietExec64:      + CategoryInfo          : OperationStopped: (Failed to create eBPF user mode service.:String) , RuntimeException
WixQuietExec64:      + FullyQualifiedErrorId : Failed to create eBPF user mode service.
WixQuietExec64:   
WixQuietExec64:  Error 0x80070001: Command line returned an error.
WixQuietExec64:  Error 0x80070001: QuietExec64 Failed
WixQuietExec64:  Error 0x80070001: Failed in ExecCommon method
CustomAction Install_Script returned actual error code 1603 (note this may not be 100% accurate if translation happened inside sandbox)
```

### Additional details

_No response_

### LINKED ISSUE #1629: MSI Installation issues
### Describe the bug

1. `netsh ebpf` command works first time when MSI is installed with default option (`runtime components`). If MSI is reused with say install the `Testing components` option to get the sample program files, `netsh ebpf` command stops working. -- already tracked by #1276
2. The ebpf binaries are left at multiple paths after installation is complete -- in `C:\Program Files\ebpf-for-windows` and `C:\Windows\System32`
3. eBPF store registry keys are not cleaned when the MSI is uninstalled.

### OS information

_No response_

### Steps taken to reproduce bug

**Issue1**
1. Install MSI with default option `runtime components`
2. `netsh ebpf` command should work
3. Re-run MSI and select `Testing components` option also.
4. `netsh ebpf` command stops working

### Expected behavior

NA

### Actual outcome

NA

### Additional details

_No response_

### PR DESCRIPTION
**Pull request by gtrevi**
Closes #1276
Closes #1412 (auto-starts drivers and service)
Closes #1478
Closes #1629
Complements work for #1883

## Description

This PR implements a new `ebpf-for-windows_installer.msi` installer, using the latest [Wix Toolset v3.11.2](https://wixtoolset.org/), and now natively integrates as a standalone and atomic project within the `ebpf-for-windows.sln` solution, without requiring the '*VC++ Redist*' anymore (ref: #1478). It therefore replaces in full the previous installer implemented using CMake scripts with `Wix` instrumentation.

The PR also adds the following missing features and improvements:

- The entire *eBPF for Windows* installation now resides within the selected install folder only (including drivers, JIT service and `netsh` helper). Previously all `.exe` and `.dll` files were also deployed to the `\Windows\System32` folder.
- Add an additional `JIT` default sub-feature for the `eBpfSvc` (parented to the `Runtime` feature), within "Debug-only" builds.
- Add automatic installation/removal of the `sample_ebpf_ext` driver, in accordance with the installation/removal of the `Testing` feature.
- With this installer, the `eBpfSvc` service (if installed with the `JIT` feature), the `eBPFCore` and `NetEbpfExt` drivers are now automatically started once the installer finalizes the installation.

From a licensing perspective, a new `LICENSE.rtf` file was added (as Wix requires a `.RTF` format) and the installer is now customized with the [approved logo](https://github.com/microsoft/ebpf-for-windows/blob/main/docs/eBPF%20logo%20png%20800px.png) from the `ebpf-for-windows` GitHub repository.

## Testing
Manual testing, on both UI & unattended command line: Install, Change Repair and Uninstall.
Installer's feature specs:

- Runtime Components: 
    - Installs core components in the `<install folder>` and `<install folder>\drivers` folders & components
    - Adds the `<install folder>` to the system's `PATH` environment variable
    - Installs & starts the `eBpfCore` driver
    - Installs & starts the `NetEbpfExt` driver
    - `netsh` extension installed and responding to `netsh ebpf`

- JIT
    - Installs the `<install folder>\JIT` folder & components
    - Installs & starts the `eBpfSvc` service

- Development Components:
    - Installs the `<install folder>\include\*` folders & files

- Testing components
    - Installs the `<install folder>\testing` folder & components
    - Installs & starts the `SampleEbpfExt` driver

## Documentation
The following documentation was updated:

- `GettingStarted.md`
- `InstallEbpf.md`
- `LICENSE.rtf`

### CODE REVIEW

File: docs/InstallEbpf.md [hunk 1]
```diff
@@ -8,20 +8,31 @@ install or update the eBPF installation in the VM.
...
-  rather than running programs, you can [use the NuGet package](GettingStarted.md#using-ebpf-in-development)
-  instead of the MSI.
-* **Testing**: this adds tests for the eBPF runtime for use by eBPF runtime developers.
+1. Download the `ebpf-for-windows_installer.msi` file from the [latest release on GitHub](https://github.com/microsoft/ebpf-for-windows/releases).
+1. Execute the `ebpf-for-windows_installer.msi` file you downloaded.
+1. After accepting the License and selecting the desired installation folder (default will be "`C:\Program Files\eBPFforWindows`"), the following components will be selectable from the *Installation Wizard*:
+
+    * **Runtime Components** (mandatory): this feature adds the eBPF runtime and core components, which are also required by the other components. If you select only this
+      feature, only [native code generation](NativeCodeGeneration.md) is enabled.
+    * **Development** (optional): this feature adds headers and libraries used for development. If you only want to use eBPF for development
```
Comment (dthaler) [inline docs/InstallEbpf.md:17]: What happened to the "JIT" feature?

File: docs/InstallEbpf.md [hunk 2]
```diff
@@ -8,20 +8,31 @@ install or update the eBPF installation in the VM.
...
+      rather than running programs, you can [use the NuGet package](GettingStarted.md#using-ebpf-in-development)
+      instead of the MSI.
+    * **Testing** (optional): this feature adds tests for the eBPF runtime for use by eBPF runtime developers.
+
+An **unattended install/uninstall** is also supported, through the following commands (examples show a full Dev&Test installation/unistallation, using "`C:\Program Files\eBPFforWindows`" as a custo installation folder):
```
Comment (dthaler) [inline docs/InstallEbpf.md:22]:
```suggestion
An **unattended install/uninstall** is also supported, through the following commands (examples show a full Dev&Test installation/uninstallation, using "`C:\Program Files\eBPFforWindows`" as a custom installation folder):
```

File: docs/InstallEbpf.md [hunk 3]
```diff
@@ -8,20 +8,31 @@ install or update the eBPF installation in the VM.
...
+
+* Installation:
+
+    ```bash
+    C:\Windows\system32\msiexec.exe /i ebpf-for-windows_installer.msi INSTALLFOLDER="C:\Program Files\eBPFforWindows" ADDLOCAL=eBPF_Development,eBPF_Testing /qn
+    ```
+    >**Note**: like in the in the graphical *Installation Wizard*, you can also customize the installation by choosing what **optional** features (i.e., `eBPF_Development` and/or `eBPF_Testing`) should be installed and assigning the comma-separated values to the `ADDLOCAL` parameter. The above commands, besides the mandatory `Runtime_Components`, also install the *Development* and *Testing* components.
```
Comment (dthaler) [inline docs/InstallEbpf.md:29]:
```suggestion
    >**Note**: like in the graphical *Installation Wizard*, you can also customize the installation by choosing what **optional** features (i.e., `eBPF_Development` and/or `eBPF_Testing`) should be installed and assigning the comma-separated values to the `ADDLOCAL` parameter. The above commands, besides the mandatory `Runtime_Components`, also install the *Development* and *Testing* components.
```

File: installer/Product.wxs [hunk 4]
```diff
@@ -0,0 +1,507 @@
...
+		<!-- Define the Product features and installation steps -->
+		<Feature Id="ProductFeature" Title="eBPF for Windows Installer" ConfigurableDirectory="INSTALLFOLDER" Display="expand" Level="1" Absent="disallow">
+			<Feature Id="eBPF_Runtime_Components" Level="1" Title="Runtime Components" Absent="disallow">
+				<ComponentGroupRef Id="eBPF_Runtime_Components" />
+				<ComponentGroupRef Id="eBPF_Service" />
+				<ComponentGroupRef Id="eBPFCore_Driver" />
+				<ComponentGroupRef Id="NetEbpfExt_Driver" />
+			</Feature>
+			<Feature Id="eBPF_Development" Level="2" Title="Development components">
+				<ComponentGroupRef Id="eBPF_Development_inlude" />
```
Comment (dthaler) [inline installer/Product.wxs:30]:
```suggestion
				<ComponentGroupRef Id="eBPF_Development_include" />
```
same typo in 31-38

File: installer/Product.wxs [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+	</Fragment>
+
+	<!-- Define the product's <Development> components -->
+	<Fragment>
+		<ComponentGroup Id="eBPF_Development_inlude" Directory="dir_include">
+			<Component Id="BPF_ENDIAN.H" DiskId="1" Guid="59B40FBC-2842-449B-9101-CFBCCF785335">
+				<File Id="BPF_ENDIAN.H" Name="bpf_endian.h" Source="..\include\bpf_endian.h" />
+			</Component>
+			<Component Id="BPF_HELPER_DEFS.H" DiskId="1" Guid="EA36437E-777C-4FFE-ADF5-D806BBDB46FE">
+				<File Id="BPF_HELPER_DEFS.H" Name="bpf_helper_defs.h" Source="..\include\bpf_helper_defs.h" />
```
Comment (dthaler) [inline installer/Product.wxs:237]: If we add another file to the include directory, it looks like we need to update this file, correct?  Any way to make this automatic in the future?

File: installer/Product.wxs [hunk 6]
```diff
@@ -0,0 +1,507 @@
...
+		</ComponentGroup>
+		<ComponentGroup Id="eBPF_Development_inlude_linux" Directory="dir_include_linux">
+			<Component Id="BPF.H_2" DiskId="1" Guid="ED4F7D03-0334-4FB3-A3AC-1DAD36FF1228">
+				<File Id="BPF.H_2" Name="bpf.h" Source="..\include\linux\bpf.h" />
+			</Component>
+			<Component Id="STDDEF.H" DiskId="1" Guid="41D2F835-B4A7-450F-B064-1280C5D55505">
+				<File Id="STDDEF.H" Name="stddef.h" Source="..\include\linux\stddef.h" />
+			</Component>
+		</ComponentGroup>
+		<ComponentGroup Id="eBPF_Development_inlude_net" Directory="dir_include_net">
```
Comment (dthaler) [inline installer/Product.wxs:332]:
```suggestion
		<ComponentGroup Id="eBPF_Development_include_net" Directory="dir_include_net">
```
make sure you search for all occurrences of "inlude" and fix them

File: installer/Product.wxs (hunk 5 above)
Comment (gtrevi) [reply installer/Product.wxs:237]: Unfortunately not (and understandably): the installer's paradigm is to *intentionally install and track components*, for reliable handling during installation and uninstallation.
What components are released with a production MSI has always to be under 100% control.

File: installer/Product.wxs (hunk 4 above)
Comment (gtrevi) [reply installer/Product.wxs]: fixed

File: installer/Product.wxs (hunk 6 above)
Comment (gtrevi) [reply installer/Product.wxs]: fixed

File: docs/InstallEbpf.md (hunk 1 above)
Comment (gtrevi) [reply docs/InstallEbpf.md:17]: @dthaler Previously, the "JIT" feature was just present in the documentation, but not in the actual MSI... therefore the documentation was documenting an inexistent MSI feature.

If there "is" a "JIT" feature that was supposed to be there, then let's sync offline and happy to package it.

Comment (dthaler) [reply docs/InstallEbpf.md:17]: See https://github.com/microsoft/ebpf-for-windows/blob/main/tools/setup/CMakeLists.txt#L90
It's basically ebpfsvc.exe.

Comment (dthaler) [reply docs/InstallEbpf.md:17]: Ok to deal with that in a separate PR though and track it as part of issue #1883.

**Review by dthaler — APPROVED** [review]

File: docs/InstallEbpf.md (hunk 1 above)
Comment (gtrevi) [reply docs/InstallEbpf.md:17]: From what I can see in the previous MSI database (and the resulting GUI), the service was installed as part of the "Runtime components":

```xml
<Component Id="CM_CP_Runtime.ebpfsvc.exe" Guid="{990D477D-5AE7-56FF-B296-5D269A9473AA}" Win64="yes">
                        <File Id="CM_FP_Runtime.ebpfsvc.exe" Name="ebpfsvc.exe" KeyPath="yes" DiskId="1" Source="SourceDir\File\CM_FP_Runtime.ebpfsvc.exe" />
</Component>
```

image


If we do need to split the service as a separate "JIT" feature, then I'd propose to just do it now OR, I can just specify more clearly in the documentation that the JIT component is among the "Runtime components".

Comment (dthaler) [reply docs/InstallEbpf.md:17]: if it's easy to do now, feel free since I think it will help with #1883.

**Review by dv-msft — APPROVED** [review]

**Review by dthaler — APPROVED** [review]

File: docs/InstallEbpf.md (hunk 1 above)
Comment (gtrevi) [reply docs/InstallEbpf.md:18]: Implemented partitioning for JIT, and available for Debug builds only.

**Comment by nibanks** [comment]
Just curious, is the goal of this installer to be both the development and runtime stuff? Or do you plan to have separate `-dev` and runtime installers? cc @mtfriesen for XDP

**Review by Alan-Jowett — DISMISSED** [review]

**Comment by gtrevi** [comment]
> Just curious, is the goal of this installer to be both the development and runtime stuff? Or do you plan to have separate `-dev` and runtime installers? cc @mtfriesen for XDP

As discussed offline, both, and it will be selectable from either the GUI or the unattended command line (and installed features will be modified accordingly).

**Review by dthaler — DISMISSED** [review]

**Review by dthaler — DISMISSED** [review]

File: installer/Product.wxs [hunk 7]
```diff
@@ -0,0 +1,561 @@
...
+				<File Source="$(var.ebpfsvc.TargetPath)" KeyPath="yes" />
+				<ServiceInstall Id="eBPFSvcInstaller" Name="eBPFSvc" DisplayName="eBPF Service" Description="eBPF Service"
+								Start="auto"
+								Type="ownProcess"
+								ErrorControl="normal"
+								Account="[WIX_ACCOUNT_LOCALSERVICE]"
+								Vital="yes"
+								Interactive="no" />
+				<ServiceControl Id="eBPFSvc_Start" Name="eBPFSvc"
+								Start="install"
```
Comment (shankarseal) [inline installer/Product.wxs:216]: I am not sure I understand. The command `eBPFSvc -Install` installs the service, not start it.

File: installer/Product.wxs [hunk 8]
```diff
@@ -0,0 +1,561 @@
...
+				<File Id="NetEbpfExt.inf" Name="NetEbpfExt.inf" Source="$(var.netebpfext.TargetDir)NetEbpfExt.inf" KeyPath="yes" />
+			</Component>
+			<Component Id="NetEbpfExt.sys" Guid="{3ADAA711-7F62-472A-A60D-589D543ED507}">
+				<File Id="NetEbpfExt.sys" Name="NetEbpfExt.sys" Source="$(var.netebpfext.TargetPath)" KeyPath="yes" />
+			</Component>
+			<Component Id="NetEbpfExt.pdb" Guid="{3B35751C-693A-47C0-A258-97741D8B72CC}">
+				<File Id="NetEbpfExt.pdb" Name="NetEbpfExt.pdb" Source="$(var.netebpfext.TargetDir)NetEbpfExt.pdb" KeyPath="yes" />
+			</Component>
+		</ComponentGroup>
+		<SetProperty Id="NetEbpfExt_Driver_install" Value='"sc.exe" create NetEbpfExt type=kernel start=demand binpath="[#NetEbpfExt.sys]"' Before="NetEbpfExt_Driver_install" Sequence="execute"/>
```
Comment (shankarseal) [inline installer/Product.wxs:265]: Can you please change this to `start= auto` as that will fix #1412 .

File: installer/Product.wxs (hunk 7 above)
Comment (gtrevi) [reply installer/Product.wxs:216]: The service is currently set to install as auto-start, i.e. `Start="auto"` within the `Serviceinstall` tag.

Setting `Start='install'`, `Stop="both"` and `Remove="uninstall"` within the `ServiceControl` tag would mean: "*start the service on install, remove the service when the product is uninstalled, and stop the service both on install and uninstall*".
`ServiceControl` is not to be confused with an `sc` ovveride.

File: installer/Product.wxs (hunk 8 above)
Comment (gtrevi) [reply installer/Product.wxs:265]: Agreed. For #1412, shall we also auto-start all the drivers, i.e. `eBpfCore` and `SampleEbpfExt` (when installed) as well?

Comment (gtrevi) [reply installer/Product.wxs]: @shankarseal I just pushed the mod to auto-start *all* the drivers, please resolve if that's what we ultimately want as spec-ed.

**Review by Alan-Jowett — APPROVED** [review]

**Pull request closed without merging**